### PR TITLE
Support settings frame

### LINKED ISSUE #775: Add support for HTTP/2 SETTINGS frame
As far as I know there is no support for the HTTP/2 SETTINGS frame (https://quicwg.github.io/base-drafts/draft-ietf-quic-http.html#rfc.section.5.2.4).
This means we cannot disable/enable fro example server push. 

Google QUIC has it:
```
[0705/020214.964850:VERBOSE1:http2_frame_decoder_adapter.cc(160)] OnFrameHeader: length=6, type=SETTINGS, flags=, stream=0
[0705/020214.964936:VERBOSE1:http2_frame_decoder_adapter.cc(374)] OnSettingsStart: length=6, type=SETTINGS, flags=, stream=0
[0705/020214.965012:VERBOSE1:http2_frame_decoder_adapter.cc(383)] OnSetting: parameter=MAX_HEADER_LIST_SIZE, value=16384
[0705/020214.965060:VERBOSE1:http2_frame_decoder_adapter.cc(398)] OnSettingsEnd
```

### PR DESCRIPTION
**Pull request by twdkeule**
Server-side support for the HTTP/2 SETTINGS frame
Required for #715 
Fixes #775

### CODE REVIEW

File: h2quic/server.go [hunk 1]
```diff
@@ -234,7 +248,30 @@ func (s *Server) handleRequest(session streamCreator, headerStream quic.Stream,
...
+	pushEnabled, ok := h2SettingsFrame.Value(http2.SettingEnablePush)
+	if ok {
+		s.pushEnabled[session.ConnectionID()] = (pushEnabled != 0)
+	}
+	// SETTINGS_HEADER_TABLE_SIZE
+	settingHeaderTableSize, ok := h2SettingsFrame.Value(http2.SettingHeaderTableSize)
+	if ok {
+		if settingHeaderTableSize != 0 {
+			// MUST be zero
+			return qerr.InternalError
```
Comment (marten-seemann) [inline h2quic/server.go]: Is there a more specific error we can return here, maybe a `qerr.InvalidHeadersStreamData`?

File: h2quic/server.go [hunk 2]
```diff
@@ -234,7 +248,30 @@ func (s *Server) handleRequest(session streamCreator, headerStream quic.Stream,
 			session.Close(nil)
 		}
 	}()
+	return nil
+}
 
+func (s *Server) handleSettingsFrame(h2SettingsFrame *http2.SettingsFrame, session streamCreator) error {
+	// PUSH
+	// TODO: in the working draft there is no EnablePush setting! This is replaced by a separate MAX_PUSH_ID frame
```
Comment (marten-seemann) [inline h2quic/server.go]: Please don't add TODOs for the IETF draft.

File: h2quic/server.go [hunk 3]
```diff
@@ -153,9 +159,17 @@ func (s *Server) handleRequest(session streamCreator, headerStream quic.Stream,
 		return qerr.Error(qerr.HeadersStreamDataDecompressFailure, "cannot read frame")
 	}
 	h2headersFrame, ok := h2frame.(*http2.HeadersFrame)
-	if !ok {
-		return qerr.Error(qerr.InvalidHeadersStreamData, "expected a header frame")
+	if ok {
```
Comment (marten-seemann) [inline h2quic/server.go]: Nit: I'd prefer a type switch here.

File: h2quic/server.go [hunk 4]
```diff
@@ -53,7 +53,8 @@ type Server struct {
 
 	supportedVersionsAsString string
 
-	pushEnabled map[protocol.ConnectionID]bool
+	pushEnabled       map[protocol.ConnectionID]bool
+	maxHeaderListSize map[protocol.ConnectionID]uint32
```
Comment (marten-seemann) [inline h2quic/server.go]: Make this a single map and introduce a `{pushEnabled bool, maxHeaderListSize uint32}` struct.

File: h2quic/server.go [hunk 5]
```diff
@@ -52,6 +52,9 @@ type Server struct {
 	listener      quic.Listener
 
 	supportedVersionsAsString string
+
+	pushEnabled       map[protocol.ConnectionID]bool
+	maxHeaderListSize map[protocol.ConnectionID]uint32
```
Comment (marten-seemann) [inline h2quic/server.go]: Only use one map here.

File: h2quic/server.go [hunk 6]
```diff
@@ -112,6 +121,7 @@ func (s *Server) serveImpl(tlsConfig *tls.Config, conn net.PacketConn) error {
 		if err != nil {
 			return err
 		}
+		s.pushEnabled[sess.ConnectionID()] = true
```
Comment (marten-seemann) [inline h2quic/server.go]: This is probably a leftover from your testing, isn't it?

File: h2quic/server_test.go [hunk 7]
```diff
@@ -325,6 +328,58 @@ var _ = Describe("H2 server", func() {
...
+		It("Disables pushes if requested", func() {
+			headerStream.dataToRead.Write([]byte{
+				// length	  |type|flags| data stream id   |
+				0x0, 0x0, 0x6, 0x4, 0x0, 0x0, 0x0, 0x0, 0x0,
+				// SETTINGS identifier			   | Value
+				0x0, uint8(http2.SettingEnablePush), 0x0, 0x0, 0x0, 0x0,
+			})
+			err := s.handleRequest(session, headerStream, &sync.Mutex{}, hpackDecoder, h2framer)
+			Expect(err).ToNot(HaveOccurred())
+			pushEnabled, ok := s.pushEnabled[session.ConnectionID()]
```
Comment (marten-seemann) [inline h2quic/server_test.go]: There's a short form: `Expect(s.pushEnabled).To(HaveKeyWithValue(session.ConnectionID(), true))`

File: h2quic/server_test.go [hunk 8]
```diff
@@ -325,6 +328,58 @@ var _ = Describe("H2 server", func() {
...
+			maxHeaderListSize := uint32(123) // must be smaller than 255 to fit in uint8 later
+			headerStream.dataToRead.Write([]byte{
+				// length	  |type|flags| data stream id   |
+				0x0, 0x0, 0x6, 0x4, 0x0, 0x0, 0x0, 0x0, 0x0,
+				// SETTINGS identifier					  | Value
+				0x0, uint8(http2.SettingMaxHeaderListSize), 0x0, 0x0, 0x0, uint8(maxHeaderListSize),
+			})
+			err := s.handleRequest(session, headerStream, &sync.Mutex{}, hpackDecoder, h2framer)
+			Expect(err).ToNot(HaveOccurred())
+			maxHeaderListSizeFromConnection, ok := s.maxHeaderListSize[session.ConnectionID()]
```
Comment (marten-seemann) [inline h2quic/server_test.go]: Use the Gomega `HaveKeyWithValue` here.

File: h2quic/server.go [hunk 9]
```diff
@@ -52,6 +52,9 @@ type Server struct {
 	listener      quic.Listener
 
 	supportedVersionsAsString string
+
+	pushEnabled       map[protocol.ConnectionID]bool
```
Comment (marten-seemann) [inline h2quic/server.go]: This map is growing indefinitely. We must delete closed connections.

**Comment by twdkeule** [comment]
I only implemted receiving the frame for now.


Is the repsonsewriter created for the entire session or for each request/response pair?

**Comment by marten-seemann** [comment]
There's one go-routine that is started for handling the header stream in https://github.com/lucas-clemente/quic-go/blob/master/h2quic/server.go#L133. For every H2 frame read from the header stream, we start another go-routine that creates the response-writer in https://github.com/lucas-clemente/quic-go/blob/master/h2quic/server.go#L202.

It would be nice if we could store the push settings as a local variable of the go-routine handling the header stream. This way, we'd avoid having a map in the `h2quic.Server`.

**Comment by twdkeule** [comment]
@marten-seemann Could you have another look?

**Comment by twdkeule** [comment]
@marten-seemann Could you have another look?

**Comment by marten-seemann** [comment]
Hi @twdkeule, how do you make Chrome send a H2 Settings Frame? I never receive one here.

**Comment by twdkeule** [comment]
To be honest, I can't get chrome to send one either.
I have a log where a Chrome/Chromium: `running on Ubuntu 16.04 Chrome/59.0.3071.109 Linux x86_64` sends a SETTINGS frame to the proto-quic toy server, that I used to develop this.
Maybe when using that specific build of Chrome would help?

**Comment by marten-seemann** [comment]
@lucas-clemente: What do you think about this?

**Comment by marten-seemann** [comment]
Closing this PR due to age.
We will need a new `h2quic` package soon anyway, with an implementation of [HTTP/3](https://quicwg.org/base-drafts/draft-ietf-quic-http.html).

**Pull request closed without merging**